File: frontend/src/api/stocks.ts
import client from './client';
import type { Page, StockRow } from './types';

// V1.5 Stage H — the stocks (股票) list endpoint.
//
// The backend reads the latest daily snapshot joined to the stock pool, with
// optional industry / tag filters, sortable columns and standard pagination.
// We surface the params as plain optional keys so callers omit what they don't
// need (axios drops undefined query params automatically).

export type StockSortKey =
  | 'pct_change'
  | 'amount'
  | 'total_mv'
  | 'pe'
  | 'price'
  | 'turnover';

// Quick-tag preset keys; must match the backend tag enum exactly.
export type StockTagKey =
  | 'limit_up'
  | 'limit_down'
  | 'top_gainers'
  | 'low_price'
  | 'high_turnover';

export interface FetchStocksParams {
  industry?: string;
  tag?: string;
  sort?: StockSortKey;
  order?: 'asc' | 'desc';
  page?: number;
  size?: number;
}

export const fetchStocks = (params: FetchStocksParams = {}) =>
  client
    .get<Page<StockRow>>('/stocks', { params })
    .then((r) => r.data);

// Distinct industries for the filter dropdown. The stocks endpoint supports an
// optional industry list helper; if absent we fall back to deriving it client
// side — but exposing it here keeps the page decoupled from that detail.
export const fetchIndustries = () =>
  client.get<string[]>('/stocks/industries').then((r) => r.data);

File: backend/app/api/stocks.py
"""Stock list router (BP-V1.5-012) — paginated browse/filter/sort.

Endpoints under ``/api/v1/stocks``.
"""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.deps import get_db
from app.services import stock_list_service

router = APIRouter(prefix="/stocks", tags=["stocks"])


def _ok(data=None, msg: str = "ok") -> dict:
    from app.main import api_ok

    return api_ok(data, msg)


@router.get("")
def list_stocks(
    industry: str | None = Query(None),
    tag: str | None = Query(
        None, pattern="^(limit_up|limit_down|top_gainers|low_price|high_turnover)$"
    ),
    sort: str = Query("pct_change", pattern="^(pct_change|amount|total_mv|pe|price|turnover)$"),
    order: str = Query("desc", pattern="^(asc|desc)$"),
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
) -> dict:
    """Paginated stock list with optional industry filter and sort."""
    return _ok(stock_list_service.list_stocks(db, industry, tag, sort, order, page, size))


@router.get("/industries")
def industries(db: Session = Depends(get_db)) -> dict:
    """Distinct industries for the filter dropdown."""
    return _ok(stock_list_service.list_industries(db))
